File: java/beer-song/src/main/java/BeerSong.java
public class BeerSong {

    static final String noMore = "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n\n";

    public String sing(int from, int num) {
        StringBuilder song = new StringBuilder();

        for(int i = from; i > from - num && i > 0; --i){
            if(i == 1){
                song.append("1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n\n");
            } else {
                song.append(i)
                    .append(" bottles of beer on the wall, ")
                    .append(i)
                    .append(" bottles of beer.\nTake one down and pass it around, ")
                    .append(i-1);
                if(i == 2)
                    song.append(" bottle of beer on the wall.\n\n");
                else
                    song.append(" bottles of beer on the wall.\n\n");
            }
        }

        if(from - num < 0)
            song.append(noMore);

        return song.toString();
    }

    public String singSong() {
        return  sing(99, 99) + noMore;
    }
}

File: csharp/beer-song/BeerSong.cs
using System;
using System.Linq;
using System.Text;

public static class BeerSong
{
    public static string Recite(int startBottles, int takeDown)
        => Enumerable.Range(startBottles - takeDown + 1, takeDown)
        .Reverse()
        .Select(Verse)
        .Aggregate("", (song, verse) => song + verse + "\n\n")
        .RemoveLast("\n\n");

    private static string Verse(int i)
        => FirstLine(i) + SecondLine(i);

    private static string RemoveLast(this string str, string rm) 
        => str.Remove(str.LastIndexOf(rm));

    private static string FirstLine(int i)
        => NumberOfBottles(i) + BottleEnding(i) + " of beer on the wall, "
            + RightPart(i) + ".\n";

    private static string SecondLine(int i)
        => ((i == 0) 
                ? "Go to the store and buy some more, " 
                : $"Take {((i == 1) ? "it" : "one")} down and pass it around, "
            )
            + RightPart((i == 0) ? 99 : i - 1) + " on the wall.";

    private static string RightPart(int i)
        => NumberOfBottles(i).ToLower() + BottleEnding(i) + " of beer";

    private static string NumberOfBottles(int i)
        => (i == 0) ? "No more" : $"{i}";

    private static string BottleEnding(int i)
        => " bottle" + ((i == 1) ? "" : "s");
}